### PR TITLE
[jit] Dict mutability

### PR DESCRIPTION
**Pull request by driazati**
Adds `aten::_set_item` for `dict[key]` calls

### CODE REVIEW

File: torch/csrc/jit/register_prim_ops.cpp [hunk 1]
```diff
@@ -1532,7 +1541,11 @@ RegisterOperators reg2({
       Operator(                                                                \
           "prim::DictIndex(Dict(" key_type ", t) self, " key_type              \
           " key) -> t",                                                        \
-          dictIndex)
+          dictIndex),                                                          \
+      Operator(                                                                \
+          "aten::_set_item(Dict(" key_type ", t)(a!) l, " key_type             \
```
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: I don't think this should return the dict as a result, it doesn't in python

File: torch/csrc/jit/register_prim_ops.cpp [hunk 2]
```diff
@@ -1532,7 +1541,11 @@ RegisterOperators reg2({
       Operator(                                                                \
           "prim::DictIndex(Dict(" key_type ", t) self, " key_type              \
```
Comment (suo) [inline torch/csrc/jit/register_prim_ops.cpp]: Now that dictionaries are mutable, all ops that extract values from the list (keys, values, index) will need to return wildcards.

**Review by suo — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/register_prim_ops.cpp (hunk 1 above)
Comment (driazati) [reply torch/csrc/jit/register_prim_ops.cpp]: This was originally just mirroring the list `_set_item`, I can't see any reason why it'd be necessary to return anything though.

**Comment by suo** [comment]
also, in alias analysis, you need to make dictionary graph inputs alias each other:
```
  for (const auto& pr : dictTypes) {
    makeAllAlias(pr.second, *aliasTracker_);
  }
```
in `analyze(Graph)`

**Comment by eellison** [comment]
> also, in alias analysis, you need to make dictionary graph inputs alias each other:
> 
> ```
>   for (const auto& pr : dictTypes) {
>     makeAllAlias(pr.second, *aliasTracker_);
>   }
> ```
> in `analyze(Graph)`

I'm not sure I fully understand - shouldn't we be making all of the input types alias each other, regardless of what type of data structure they are contained in? 

Doesn't the following example fail:

@torch.jit.script
def test(a: Tuple[Tensor], b: List[Tensor]):
     ....

ten = torch.tensor(1.0)
tup = (ten,)
ten_list = [ten]
test(tup, ten_list)

**Comment by suo** [comment]
Right now all elements extracted from a list/dict are wildcards so they're aliased anyway. Once we improve that granularity, we'll need to revisit.

**Review by suo — APPROVED** [review]

**Pull request closed without merging**